feat(radio): support aria-label(ledby) on md-radio

MartinMa opened the pull request:
Adds feature #586 
**radio: adding actionName attribute as label fallback (aria-label)**

Added

- [x] tests
- [x] revised the documentation
- [x] and improved the examples/demo.

Please give me feedback, in case something needs to be changed.

DevVersion commented on src/components/radio/README.md:
```diff
@@ -54,6 +54,7 @@ The `md-radio-group` component has no button initially selected.
 | `value` | `any` | The value of this radio button. |
 | `checked` | `boolean` | Whether the radio is checked. |
 | `disabled` | `boolean` | Whether the radio is disabled. |
+| `actionName` | `string` | Used to set the `aria-label` attribute of the underlying input element. |
```
This should be `aria-label`. We just want to forward the attribute. 

- As same as in the slide-toggle

MartinMa replied:
I think it is not a good idea to have a *pseudo* `aria-label` attribute on the MdRadio component and at the same time a real one on the underlying input element. It might confuse screen readers and its users. Do you know what I mean? That's why I chose a different name as seen on the [Angular 2 documentation for template syntax](https://angular.io/docs/ts/latest/guide/template-syntax.html) (see Property Binding / We must use attribute binding when there is no element property to bind).

But while taking a look a the current implementation for MdSlideToggle, I see that the property used there is named `ariaLabel`, which is different from `aria-label`. So this subtle difference is probably enough avoid potential issues with screen readers.

So if everyone agrees and doesn't see any more issues, I can change the property name (and add `aria-labelledby` as well).

DevVersion replied:
We also using the `aria-label` input in the checkbox as well. (See [here](https://github.com/angular/material2/blob/master/src/components/checkbox/checkbox.ts#L80)).

Screenreaders shouldn't be confused about that attribute, because the host element is normally not focusable.

Without any attribute binding, you're still using the dash cased attribute
```html
<md-radio-button aria-label="My Label Input">
```

But with attribute binding, you're right, it will use the camelCased attribute name.

MartinMa replied:
Thanks for the explanation. I will conform to the pattern used for the checkbox component.

DevVersion commented on src/components/radio/radio.html:
```diff
@@ -13,6 +13,7 @@
           [checked]="checked"
           [disabled]="disabled"
           [name]="name"
+          [attr.aria-label]="actionName"
```
We should also support the `aria-labelled-by` attribute.

MartinMa replied:
Agreed.

DevVersion commented on src/components/radio/radio.spec.ts:
```diff
@@ -393,6 +401,20 @@ describe('MdRadio', () => {
+      let nativeRadioInput = <HTMLElement> fruitRadioNativeElements[1].querySelector('input');
+
+      fixture.detectChanges();
+      expect(nativeRadioInput.getAttribute('aria-label')).toBe(null);
```
I would also test the `aria-label` and `aria-labelled-by` attribute as a bound property, which means modifying the `@Input`'s input at runtime triggering a change detection cycle.

DevVersion commented on src/demo-app/radio/radio-demo.html:
```diff
@@ -28,3 +28,30 @@
   </md-radio-group>
   <p>Your favorite season is: {{favoriteSeason}}</p>
 </section>
+<h1>ARIA Example</h1>
```
Not sure about adding an extra demo, just for the `aria-label` and `aria-labelled-by` attribute. It seems unnecessary to me.

jelbourn commented:
Agreed on all of @DevVersion's comments

MartinMa commented:
I updated the pull request.

- [x] Changed property to `ariaLabel` (corresponding attribute is called `aria-label`).
- [x] Added property `ariaLabelledby` (corresponding attribute is called `aria-labelledby`).
- [x] Updated tests and added new tests.
- [x] Updated README.
- [x] Removed demo/example code.

Let me know, if this is ok. I will squash all commits then, so it will be ready to be merged.

jelbourn commented on src/components/radio/radio.spec.ts:
```diff
@@ -393,6 +409,40 @@ describe('MdRadio', () => {
+    });
+
+    it('should not add aria-label attribute if not defined', () => {
+      expect(fruitRadioNativeInputs[1].getAttribute('aria-label')).toBe(null);
```
I'd use `hasAttribute` here and in other tests where you're checking for the _presence_ of the attribute.

jelbourn commented:
Looks good aside from that one minor comment

MartinMa commented:
Thanks for the feedback. I updated the two tests to use `hasAttribute` and squashed all commits. :hammer:

jelbourn commented:
LGTM

jelbourn commented:
Thanks!
